enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,107 @@
+---
+title: upgrade-preflight-checks-for-operators
+authors:
+  - "@LalatenduMohanty"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-06-04
+last-updated: 2020-06-05
+status: implementable
+---
+
+# Upgrade Preflight Checks For Operators
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+* How CVO can get the list of operators which has the preflight checks implemented?
+* What information CVO needs from the operators to create the preflight jobs?
+* Does CVO already has capability to create jobs for preflight checks?
+* How to trigger execution of preflight checks?
+* How CVO will report when it gets the result from preflight checks?
```

Comment:
This could go on a new property under `Release`, if we want to build on #123.  Or it could be a new, independent custom resource.  The trade-offs are a wish to limit ClusterVersion growth vs. managing preflight result removal.

Managing removal doesn't seem that bad; you could use logic like:

1. Fetch a current ClusterVersion and extract its `status.availableUpdates`.
2. Iterate over preflight-result resources.  For each resource:
    1. If its release pullspec is not in `availableUpdates`, delete the resource.
    2. If its release pullspec is in `availableUpdates`...
        1. ... and the last-audited timestamp is recent enough, skip it this round.
        2. ... and the last-audited timestamp is not recent enough, trigger a Job to update the preflight-result resource.

So I'm currently in favor of a new custom resource and not in favor of extending `Release`.

Reply:
Big +1 for a new custom resource.

Reply:
Sounds good. A new custom resource sounds fine to me.

Reply:
Fixed in https://github.com/openshift/enhancements/commit/4cadad51eb7b2e999146ae9e8af76e57f9c90556#diff-f53e26d63b28b30cadb2bd5bc6342812R80